File: Leetcode/dfs_recursion_backtracking/17_LetterCombinationsOfaPhoneNumber/Solution.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        dic = {1:"", 2:"abc", 3:"def", 4:"ghi", 5:"jkl",
               6:"mno", 7:"pqrs", 8:"tuv", 9:"wxyz"}
        res = []
        def dfs(temp, index):
            if index >= len(digits):
                if temp:
                    res.append("".join(temp))
                return
            k = int(digits[index])
            for ch in dic[k]:
                dfs(temp+[ch], index+1)
        dfs([], 0)
        return res

File: Leetcode/dfs_recursion_backtracking/17_LetterCombinationsOfaPhoneNumber/Solution.java
class Solution {
    static final HashMap<Integer, String> map = new HashMap<Integer, String>(){{
        put(2, "abc"); put(3, "def"); put(4, "ghi"); put(5, "jkl");
        put(6, "mno"); put(7, "pqrs"); put(8, "tuv"); put(9, "wxyz");
    }};
    public List<String> letterCombinations(String digits) {
        LinkedList<String> res = new LinkedList<>();
        dfs(res, digits, "", 0);
        return res;
    }
    private void dfs(LinkedList<String> res, String digits,
                     String temp, int index){
        if(index>=digits.length()){
            if(!temp.isEmpty()){
                res.add(temp);
            }
            return;
        }
        int k = Integer.parseInt( Character.toString(digits.charAt(index)) );
        for(Character c: map.get(k).toCharArray()){
            dfs(res, digits, temp+c, index+1);
        }
    }
}
